### PR TITLE
Search branch

### PR DESCRIPTION
**Pull request by mirashafran**
This pull request introduces new server-side and client-side features to enhance the search functionality, including autocomplete and result display.

Server-Side:

**API Endpoints:**
- Added a new API endpoint for searching content.
- Implemented another API endpoint for search suggestions to support the autocomplete functionality.

Added a dedicated page for handling the search method logic.

Client-Side:

components:
- Search Line: A component for the search input field.
- Search Autocomplete: Handles the display of search suggestions while typing.
- Search Results: Displays the results of the search query.

### CODE REVIEW

File: client/src/components/AutoComplete/autocompleteComponent.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        setInputValue(value);
+        setQuery(value);
+
+        if (value.length === 0) {
+            // reset suggestions when query is empty
+            setSuggestions([]);
+        } else if(value.length > 2) {
+            try {
+                const response = await axios.post('http://localhost:5000/api/content/search/suggestions', { query: value });
+                setSuggestions(response.data);
```
Comment (noaazerrad) [inline client/src/components/AutoComplete/autocompleteComponent.js]: check if you need to parse the response data? is it json?

File: client/src/components/Results/searchResultsComponent.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
+import React from 'react';
+import styles from './searchResults.module.css'
+
+/*const handleResultClick = (result) => {
+  console.log("Result clicked: ", result);
```
Comment (noaazerrad) [inline client/src/components/Results/searchResultsComponent.js]: don't keep commented code, its better to delete it if its not needed

File: client/src/components/AutoComplete/autocompleteComponent.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+    const handleSuggestions = async (e) => {
+        const value = e.target.value
+        setInputValue(value);
+        setQuery(value);
+
+        if (value.length === 0) {
+            // reset suggestions when query is empty
+            setSuggestions([]);
+        } else if(value.length > 2) {
```
Comment (noaazerrad) [inline client/src/components/AutoComplete/autocompleteComponent.js]: what if the length is eq to 2 or 1 ?

File: client/src/components/Results/searchResultsComponent.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
+import React from 'react';
+import styles from './searchResults.module.css'
+
+/*const handleResultClick = (result) => {
+  console.log("Result clicked: ", result); 
+
+}*/
+
+const SearchResults = ({ results, searchPerformed}) => {
+  if (!searchPerformed || results.length === 0) {
```
Comment (noaazerrad) [inline client/src/components/Results/searchResultsComponent.js]: why do you need both variables? why not just use the results.length?

File: server/controllers/contentController [hunk 5]
```diff
@@ -21,6 +21,49 @@ const createContent = async (req, res) => {
     }
 }
 
+// Search for content
```
Comment (noaazerrad) [inline server/controllers/contentController]: your are missing .js in the file name (currently its just a plain text file)

File: server/controllers/contentController [hunk 6]
```diff
@@ -21,6 +21,49 @@ const createContent = async (req, res) => {
...
+    res.status(200).json(results);
+
+    } catch (err) {
+        res.status(500).json({mssg: 'Error searching content', err});
+    }
+
+};
+
+// Search suggestions for autocomplete
+const searchSuggestions = async (req, res) => {
```
Comment (noaazerrad) [inline server/controllers/contentController]: i'm a bit confused regarding the naming of this action - is it suggestions or autocomplete? (in the client the component name is autocomplete) basically these are two different features: one autocomplete the text and the other returns suggestions based on the input

File: server/controllers/contentController [hunk 7]
```diff
@@ -21,6 +21,49 @@ const createContent = async (req, res) => {
...
+    const {query} = req.body; 
+    
+    try{
+        const suggestions = await content.find(
```
Comment (noaazerrad) [inline server/controllers/contentController]: also - you are allowing to search by multiple fields but presenting suggestions only for one field, this behaviour might be confusing for the user.

**Review by noaazerrad — APPROVED** [review]

File: client/src/components/Search/searchComponent.js [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+        onKeyDown={handleKeyDown}
+        className={styles.searchInput}
+
+      />
+
+      <FaSearch className={styles.searchIcon} onClick={handleSearch} />
+      {query && <FaTimes className={styles.clearIcon} onClick={clearSearch} />}
+
+      </div>
+      <Suggestions setQuery={setQuery} suggestions={suggestions} setSuggestions={setSuggestions} />
```
Comment (noaazerrad) [inline client/src/components/Search/searchComponent.js:70]: if a user clicks on the suggestion , how is the search request is triggered?

File: client/src/components/Search/searchComponent.js [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+      <SearchResults results={results} searchPerformed={searchPerformed}  />
```
Comment (noaazerrad) [inline client/src/components/Search/searchComponent.js:71]: the results should be clickable? (is it dependent on rotem's feature?)

File: client/src/components/Search/searchComponent.js (hunk 8 above)
Comment (mirashafran) [reply client/src/components/Search/searchComponent.js:70]: When a user clicks on a suggestion it completes the name of the video in the search bar and then the user can click on "search"

**Pull request merged**